### PR TITLE
SNOW-976704 : Adding lead and lag dataframe function

### PR DESCRIPTION
**Pull request by sfc-gh-rsureshbabu**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

[SNOW-976704](https://snowflakecomputing.atlassian.net/browse/SNOW-976704)

The PR is adding the lead and lag  functions as proposed in https://docs.google.com/document/d/14J5lr_a3fE1xeU-YzKWdeduBdC_035sry11lJUjjkT8/edit

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

[SNOW-976704]: https://snowflakecomputing.atlassian.net/browse/SNOW-976704?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/snowflake/snowpark/dataframe_transform_functions.py [hunk 1]
```diff
@@ -120,3 +122,85 @@ def moving_agg(
                     agg_df = agg_df.with_column(formatted_col_name, agg_col)
 
         return agg_df
+
+    def compute_lag(
```
Comment (sfc-gh-wezhou) [inline src/snowflake/snowpark/dataframe_transform_functions.py:126]: looks like compute_lag and compute_lead shares a lot common code? can we extract the common part?

Comment (sfc-gh-rsureshbabu) [reply src/snowflake/snowpark/dataframe_transform_functions.py:157]: done.

File: src/snowflake/snowpark/dataframe_transform_functions.py [hunk 2]
```diff
@@ -57,6 +59,37 @@ def _validate_formatter_argument(self, fromatter):
...
 
+    def _compute_window_function(
+        self,
+        cols: List[Union[str, Column]],
+        periods: List[int],
+        order_by: List[str],
+        group_by: List[str],
+        col_formatter: Callable[[str, str, int], str],
+        window_func: Callable[[Column, int], Column],
+        func_name: str,  # Either "LAG" or "LEAD"
```
Comment (sfc-gh-wezhou) [inline src/snowflake/snowpark/dataframe_transform_functions.py:70]: nit: move the comment to line 73?

File: src/snowflake/snowpark/dataframe_transform_functions.py [hunk 3]
```diff
@@ -57,6 +59,37 @@ def _validate_formatter_argument(self, fromatter):
...
+    ) -> "snowflake.snowpark.dataframe.DataFrame":
+        """
+        Generic function to create window function columns (lag or lead) for the DataFrame.
+        """
+        self._validate_string_list_argument(order_by, "order_by")
+        self._validate_string_list_argument(group_by, "group_by")
+        self._validate_positive_integer_list_argument(periods, func_name.lower() + "s")
```
Comment (sfc-gh-tbao) [inline src/snowflake/snowpark/dataframe_transform_functions.py:77]: QQ: why do you need append s on function name?

**Review by sfc-gh-wezhou — APPROVED** [review]

**Review by sfc-gh-sfan — COMMENTED** [review]
Let's add some doc test to serve as example

File: src/snowflake/snowpark/dataframe_transform_functions.py (hunk 3 above)
Comment (sfc-gh-rsureshbabu) [reply src/snowflake/snowpark/dataframe_transform_functions.py:77]: The argument name is "lags" for the lag function and "leads" for the lead function.

File: src/snowflake/snowpark/dataframe_analytics_functions.py [hunk 4]
```diff
@@ -85,6 +87,39 @@ def _validate_formatter_argument(self, fromatter):
...
+        window_spec = Window.partition_by(group_by).order_by(order_by)
+        df = self._df
+        for c in cols:
+            for period in periods:
+                column = _to_col_if_str(c, f"transform.compute_{func_name.lower()}")
+                window_col = window_func(column, period).over(window_spec)
+                formatted_col_name = col_formatter(
+                    column.get_name().replace('"', ""), func_name, period
+                )
+                df = df.with_column(formatted_col_name, window_col)
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/dataframe_analytics_functions.py:119]: It would be more optimal to use `df.with_columns()`.

Comment (sfc-gh-rsureshbabu) [reply src/snowflake/snowpark/dataframe_analytics_functions.py]: changed it.

**Review by sfc-gh-yixie — APPROVED** [review]

**Pull request merged**